### PR TITLE
Auth token issues solver. [DO NOT MERGE]

### PR DESCRIPTION
**Pull request by diegoreymendez**
Takes care points 3, 4 and 5 from [this issue](https://github.com/wordpress-mobile/WordPress-iOS/issues/3964) reported by @astralbodies.

This should NOT be merged until [this PR](https://github.com/wordpress-mobile/WordPress-iOS/pull/3956) by @astralbodies is merged.  The reason is that this PR is a branch from that PR.

Please keep in mind this is a hotfix, so even though some code has been moved, as little as possible of the existing logic has been really modified.  This means there's some code duplication.

**Test 1:**
- Comment out method `setUsername:` in `WPAccount`
- Login with a mixed case WP.com username name or a WP.com email address
- Create post, quit app without hitting the save button
- Reopen the app, it should ask you to re-authenticate (if you disabled the method above)
- Re-authenticate
- Make sure your post is still there

**Test 2:**
- Comment out method `setUsername:` in `WPAccount`
- Login with a mixed case WP.com username name or a WP.com email address
- Go to the site picker and add self-hosted blog
- Quit the app normally
- Reopen the app, it should ask you to re-authenticate (if you disabled the method above)
- Hit cancel instead of re-authing
- A warning will ask if you want to delete your local wp.com data
- Delete it
- Make sure the self hosted blog is still there, but that the wp.com related data has been removed (such as all wp.com blogs from the site picker).

**Test 3:**
- Re-enable method `setUsername:` in `WPAccount`
- Login with either email or mixed cap username
- Quit app, reopen
- Should log in normally, without issues

**Test 4:** 
- Comment out method `setUsername:` in `WPAccount`
- Login with a mixed case WP.com username name or a WP.com email address
- Quit the app normally
- Reopen the app, it should ask you to re-authenticate (if you disabled the method above)
- Try logging-in with a bad password.
- Make sure you can interact with the overlay that comes up, and dismiss it by pressing "OK".

**Test 5:**
- Make sure to repeat all previous tests for different iOS versions.

**Discussion:** (please take a moment to read and provide feedback)
- We don't currently have any procedure to make sure the user re-authenticates with the same user he was using previously, which seems a bit dangerous to me not to have.  This feels like something doable, but completely outside of the scope of a simple hot-fix (especially a nightly hot-fix like this one).  We could enhance `LoginViewController` with code to check that `account.userId` matches the remote info returned by the new authentication attempt.
- I believe we could improve this PR **drastically* by adding an automatic recovery attempt.  I've reviewed the keychain code we have and it supports case insensitive searches and returning multiple results.  If the auth token is `nil` we could try to (case-insensitively) search the keychain for all `account.username` and `account.email` matches, then try logging in with the returned authTokens and, on success, confirming that our locally stored `account.userId` matches the account data returned by the services (in order to make sure we're logging into the right account).  If the automatic attempt fails we can fall back into this code.

### CODE REVIEW

**Comment by diegoreymendez** [comment]
I need as many eyes as possible on this @sendhil, @SergioEstevao, @astralbodies, @koke, @jleandroperez, @kwonye, @aerych.  Thank you all in advance, guys!

**Comment by diegoreymendez** [comment]
PS: Even though I've tested this thoroughly it's a bit late here, so please make sure to check the code as much as possible.  Also if possible, review the english text for the alert dialogues I've added.

File: WordPress/Classes/Utility/WPAuthTokenIssueSolver.h [hunk 1]
```diff
@@ -0,0 +1,24 @@
+#import <Foundation/Foundation.h>
+
+typedef void(^WPAuthTokenissueSolverCompletionBlock)();
```
Comment (jleandroperez) [inline WordPress/Classes/Utility/WPAuthTokenIssueSolver.h]: You're seriously gonna love Swift!

**Comment by jleandroperez** [comment]
Some incoming feedback!

**Test 1: Mixed Case Scenario**
I couldn't trigger the lost token bug with a WordPress.com Username with mixed case (and confirming, with the *setUsername* method commented out. 

**Test 1: Email Scenario**
I'm getting reauthenticated, but when i'm back at the app, the post contents are gone. Sample [video available here](https://cloudup.com/cvqIlXGJjLn)

File: WordPress/Classes/Utility/WPAuthTokenIssueSolver.m [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+ *
+ *  @param      okBlock     The block that will be executed if the user confirms the operation.
+ */
+- (void)showCancelReAuthenticationAlertAndOnOK:(WPAuthTokenissueSolverCompletionBlock)okBlock
+{
+    NSParameterAssert(okBlock);
+    
+    NSString *alertTitle = NSLocalizedString(@"Careful!",
+                                             @"Title for the warning shown to the user when he refuses to re-login when the authToken is missing.");
+    NSString *alertMessage = NSLocalizedString(@"If you proceed, all your WP.com account data will be removed from this device.  This means any unsaved data such as locally stored posts will be deleted.  This will not affect data that has already been uploaded to your WP.com account.",
```
Comment (aerych) [inline WordPress/Classes/Utility/WPAuthTokenIssueSolver.m]: We should spell out WordPress.com. Also maybe tweak the message to be a bit more friendly and less technical. Maybe something like:

> Proceeding will remove all WordPress.com data from this device, and delete any locally saved drafts.  You will not loose anything already saved to your WordPress.com blog(s).

File: WordPress/Classes/Utility/WPAuthTokenIssueSolver.m [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+}
+
+/**
+ *  @brief      Shows the alert explaining the authToken issue to the user.
+ */
+- (void)showExplanationAlertForReAuthenticationDueToMissingAuthToken
+{
+    NSString *alertTitle = NSLocalizedString(@"Oops!",
+                                             @"Title for the warning shown to the user when the app realizes there should be an auth token but there isn't one.");
+    NSString *alertMessage = NSLocalizedString(@"We have detected a synchronization problem with your locally stored WP.com credentials. You're going to be prompted to re-authenticate.",
```
Comment (aerych) [inline WordPress/Classes/Utility/WPAuthTokenIssueSolver.m]: Also need to spell out WordPress.com here and probably need to reduce the technical level of the message.  Maybe something very simple like: 

> There was a problem connecting to WordPress.com.  Please sign in again.

**Comment by jleandroperez** [comment]
**Update:**
All test passed (i managed to trigger the bug with a mixed case username, finally).
Only pending is State Restoration, which seems not to be working for me.

Is it possible that it was actually broken in 5.3? (See #3742)

Great job Diego!

**Comment by aerych** [comment]
Hi Diego!  Testing Feedback for ya:

General: I can only trigger the bug when logging in with an email address. I haven't been able to trigger it with a mixed case username. 

Test 1: Pass
Test 2: Pass
Test 3: Pass
Test 4: Pass

File: WordPress/Classes/Utility/WPAuthTokenIssueSolver.m (hunk 3 above)
Comment (diegoreymendez) [inline WordPress/Classes/Utility/WPAuthTokenIssueSolver.m]: :+1:  Thanks!

File: WordPress/Classes/Utility/WPAuthTokenIssueSolver.m (hunk 2 above)
Comment (diegoreymendez) [inline WordPress/Classes/Utility/WPAuthTokenIssueSolver.m]: :+1: Thank you, Sir!

File: WordPress/Classes/Utility/WPAuthTokenIssueSolver.h (hunk 1 above)
Comment (diegoreymendez) [inline WordPress/Classes/Utility/WPAuthTokenIssueSolver.h]: Why's that?

**Comment by kwonye** [comment]
For Test 1, when you say quitting the app without clicking the save button, you mean drafting up a post and then quitting the app right? If so, when I log back in, the post is not there. Also, when I go to the site picker it crashes.

All other tests pass on iOS 8.4 iPhone 6.

**Comment by diegoreymendez** [comment]
@kwonye - I mean these steps:

- Open up the editor
- Type title
- Type contents
- While the editor is still open, quit the app as you would normally do on your phone (don't stop xcode).

When you reopen the app the post should NOT be immediately visible, nor available when you press the "New post" button in the nav bar.  It should be available in the list of posts for the blog where it was being created though.

Do you think you can make a video showing the crash situation @kwonye?  I haven't been able to get that and have been testing quite a bit on my end.

Thanks for the help!

**Comment by jleandroperez** [comment]
**Update Mark II:**
As @diegoreymendez pointed out... if you tap over the `New Post` button, the app is expected to create a new post.

What we actually have to do is open the Post List for the blog in which we were editing, and the draft will show up there.

Note there: WordPress.com posts were recovered, while self hosted seem to be loosing posts during this reauth process.

**Comment by diegoreymendez** [comment]
Thanks @jleandroperez - I'll check the self hosted issue to see if I can fix that.

File: WordPress/Classes/System/WordPressAppDelegate.m [hunk 4]
```diff
@@ -114,69 +117,27 @@ - (BOOL)application:(UIApplication *)application willFinishLaunchingWithOptions:
 {
     [WordPressAppDelegate fixKeychainAccess];
 
+    // Basic networking setup
+    [self setupReachability];
+    
+    // Set the main window up
+    CGRect bounds = [[UIScreen mainScreen] bounds];
```
Comment (sendhil) [inline WordPress/Classes/System/WordPressAppDelegate.m]: I think some of these lines might not be necessary for now as as a result of https://github.com/wordpress-mobile/WordPress-iOS/pull/3937, but we can keep them for now and just patch them out later. No need to introduce anything else new with this.

File: WordPress/Classes/Utility/WPAuthTokenIssueSolver.h (hunk 1 above)
Comment (sendhil) [inline WordPress/Classes/Utility/WPAuthTokenIssueSolver.h]: `() -> ()`

Comment (sendhil) [inline WordPress/Classes/Utility/WPAuthTokenIssueSolver.h]: Little nitpick, probably want to capitalize the `i` in `issue` so it's `WPAuthTokenIssueSolverCompletionBlock`

File: WordPress/Classes/System/WordPressAppDelegate.m (hunk 4 above)
Comment (diegoreymendez) [inline WordPress/Classes/System/WordPressAppDelegate.m]: :+1:

**Comment by sendhil** [comment]
So one way to test this out without commenting out `setUsername:` is:
1) Checkout `release/5.3`
2) Login with an email or a capitalized username.
3) Wait about 5 seconds.
4) Force close the app
5) Open the app up to test the app crashing.
6) Switch over to this branch
7) Run it

p.s. `git checkout -` is your friend.

**Comment by sendhil** [comment]
oh, also if you are testing this out, remember to reset your simulator before you start otherwise there are some ghosts of keychain past that might suppress this bug.

**Comment by sendhil** [comment]
Just to mirror my comments from Slack, I got all the test conditions to work except the second one as with the second one my notifications were still intact.

File: WordPress/Classes/System/WordPressAppDelegate.m [hunk 5]
```diff
@@ -998,6 +1027,15 @@ - (void)removeTodayWidgetConfiguration
     [service removeTodayWidgetConfiguration];
 }
 
+#pragma mark - Simperium helpers
+
+- (void)resetSimperiumOnAuthTokenIssue
```
Comment (sendhil) [inline WordPress/Classes/System/WordPressAppDelegate.m]: Maybe leave a note on here for people to not call this and that this was specifically made for a hotfix?

**Comment by diegoreymendez** [comment]
Fixed the notifications issue, also added the comments to the new temporary notifications method.

**Comment by sendhil** [comment]
:shipit:

File: WordPress/Classes/System/WordPressAppDelegate.m (hunk 4 above)
Comment (koke) [inline WordPress/Classes/System/WordPressAppDelegate.m]: It seems those lines made it back to develop, reopening #1162  :(

Comment (sendhil) [inline WordPress/Classes/System/WordPressAppDelegate.m]: Doh, sorry @koke

**Pull request merged**